### PR TITLE
Enable KafkaSource scale subresource

### LINKED ISSUE #137: [KafkaSource] Enable scaling via scale subresource
**Problem**
The eventing autoscaler is currently tightly-coupled with the KafkaSource implementation, making both components difficult to update. Instead the eventing autoscaler should only be interacting with the KafkaSource API

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)**
Which persona is this feature for?

**Exit Criteria**
A measurable (binary) test that would indicate that the problem has been resolved.
`k scale` should work on KafkaSource

**Time Estimate (optional):**
How many developer-days do you think this may take to resolve?
~1

**Additional context (optional)**
Add any other context about the feature request here.

https://github.com/knative-sandbox/eventing-autoscaler-keda/issues/26

### PR DESCRIPTION
**Pull request by lionelvillard**
Fixes #137

## Proposed Changes

- Enable KafkaSource scale subresource to scale the underlying deployment
-
-

**Release Note**

```release-note
🎁 The KafkaSource scale subresource can now be used to scale up and down the underlying deployment
```

**Docs**

Depends on https://github.com/knative/pkg/pull/1833 

~I'll remove the dependency on my fork my the PR above is merged (or another PR handling the scale subresource properly)~. Done

FYI @zroubalik. This PR breaks eventing Keda integration. I'm working on a PR to fix it.

### CODE REVIEW

File: config/source/resources/kafkasource.yaml [hunk 1]
```diff
@@ -48,6 +48,11 @@ spec:
   scope: Namespaced
   subresources:
     status: {}
+    scale:
+      # specReplicasPath defines the JSONPath inside of a custom resource that corresponds to Scale.Spec.Replicas.
+      specReplicasPath: .spec.replicas
+      # statusReplicasPath defines the JSONPath inside of a custom resource that corresponds to Scale.Status.Replicas.
+      statusReplicasPath: .status.replicas
```
Comment (zroubalik) [inline config/source/resources/kafkasource.yaml:55]: Are `replicas` defined on `KafkaSourceStatus` ?

Comment (lionelvillard) [reply config/source/resources/kafkasource.yaml:55]: Not yet. Good catch!!

**Comment by lionelvillard** [comment]
/hold

Let's wait for the scheduled dependency upgrade to pick up knative/pkg#1833

**Comment by lionelvillard** [comment]
@zroubalik @matzew do you mind looking at this one?

File: pkg/apis/sources/v1alpha1/kafka_types.go [hunk 2]
```diff
@@ -121,6 +127,11 @@ type KafkaSourceStatus struct {
 	// * SinkURI - the current active sink URI that has been configured for the
 	//   Source.
 	duckv1.SourceStatus `json:",inline"`
+
+	// Total number of non-terminated replicas targeted by this source
+	// +optional
+	// For round-tripping only
+	Replicas int32 `json:"replicas,omitempty"`
```
Comment (matzew) [inline pkg/apis/sources/v1alpha1/kafka_types.go:134]: Is that causing some "migration" problems?

I know we do lack "upgrade tests" for the kafka stuff

File: pkg/apis/sources/v1beta1/kafka_types.go [hunk 3]
```diff
@@ -67,6 +67,11 @@ type KafkaResourceSpec struct {
 
 // KafkaSourceSpec defines the desired state of the KafkaSource.
 type KafkaSourceSpec struct {
+	// Number of desired replicas. This is a pointer to distinguish between explicit
+	// zero and not specified. Defaults to 1.
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
```
Comment (matzew) [inline pkg/apis/sources/v1beta1/kafka_types.go:73]: Doing it on `v1`? instead ? :grimacing:

Comment (lionelvillard) [reply pkg/apis/sources/v1beta1/kafka_types.go:73]: We should allow adding new fields to an existing API without changing the version. Kubernetes allows it. See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-unstable-features-to-stable-versions

Excerpt: `.. It is generally allowed to add new fields without changing the API version, ..`

File: pkg/apis/sources/v1alpha1/kafka_types.go (hunk 2 above)
Comment (lionelvillard) [reply pkg/apis/sources/v1alpha1/kafka_types.go:134]: I don't think so since this is a new field

Comment (zroubalik) [reply pkg/apis/sources/v1alpha1/kafka_types.go:134]: Does it make a sense to move this to `duckv1.SourceStatus`?

Comment (lionelvillard) [reply pkg/apis/sources/v1alpha1/kafka_types.go:134]: I think so, as an optional field. From a conformance PoV, a source *may* support horizontal scaling. The counter argument is we tend to only declare fields that are used. Not all sources supports `replicas`. 

If we decide to add it to `duckv1.SourceStatus` later on, it won't be breaking the API so I suggest we leave it this way for now.

**Review by slinkydeveloper — COMMENTED** [review]
Does that field makes sense to exist when `KafkaSource` will be implemented with eventing-kafka-broker data plane? @pierDipi 

My feeling (i might be wrong) is that it kinda creates a "stronger" bond between the implementation, which today single tenant (vs an eventual one with eventing-kafka-broker **multi-tenant**). What this `replicas` means in a multi tenant impl?

**Comment by pierDipi** [comment]
I guess it means multiple consumers with the same consumer group.

**Comment by slinkydeveloper** [comment]
> I guess it means multiple consumers with the same consumer group.

Technically a single instance can have multiple consumers in the same consumer group, in java this is done automatically by the kafka client

**Comment by slinkydeveloper** [comment]
What I'm trying to say is that my feeling is that this field leaks an implementation detail of this particular implementation of `KafkaSource`

**Comment by pierDipi** [comment]
Can you explain a little bit more? 
I mean in a multi-tenant scenario replicas means run N consumers with the same consumer group where they run I don't care.

**Comment by slinkydeveloper** [comment]
If `replicas` means `consumers`, then why naming it `replicas`? My understanding is that this proposal wants to add `replicas`, in the meaning of the kubernetes replicas for a specific deployment

**Comment by lionelvillard** [comment]
The actual definition of `replicas` depends on the implementation. For the single-tenant KafkaSource, it really means N pods replicas. For a simple multi-tenant KafkaSource implementation it means run N consumers (by lowering throttling). 

Maybe `replicas` is not the right name. In the design document I call it `virtual replicas`.

**Comment by slinkydeveloper** [comment]
Then what about `consumers` to map with the concepts of kafka?

**Comment by lionelvillard** [comment]
+1 for `consumers`. 

The only caveat is a multi tenant source *might* provide more *virtual* consumers than the number of partitions, potentially leading to some confusion (source says 10 consumers but there is only 5 partitions in the topic). The documentation should leave the door open to such implementation.

**Comment by lionelvillard** [comment]
@travis-minke-sap @eric-sap do you have any comments/objections on this PR? (sorry I meant to ping you earlier on this one).

**Comment by pierDipi** [comment]
It seems that more consumers than the number of partitions are allowed by this implementation as well.

**Comment by lionelvillard** [comment]
> It seems that more consumers than the number of partitions are allowed by this implementation as well.

that's right. Currently only KEDA makes sure that the number of consumers does not go over. This is a problem especially if consumers are virtual consumers. I'm working on a PoC right now to give more meat to the discussion.

File: pkg/apis/sources/v1alpha1/kafka_types.go [hunk 4]
```diff
@@ -95,6 +95,17 @@ type KafkaSourceSpec struct {
 	// +optional
 	// Needed for supporting round-tripping
 	CloudEventOverrides *duckv1.CloudEventOverrides `json:"ceOverrides,omitempty"`
+
+	// Number of desired consumers. Defaults to 1.
+	//
+	// A consumer is an unit for measuring resource allocation,
+	// for example the number of pods, or the allocated bandwidth.
```
Comment (slinkydeveloper) [inline pkg/apis/sources/v1alpha1/kafka_types.go:102]: The allocated bandwidth?

What about "the number of consumers running in the consumer group"?

File: pkg/apis/sources/v1alpha1/kafka_types.go [hunk 5]
```diff
@@ -121,6 +132,11 @@ type KafkaSourceStatus struct {
 	// * SinkURI - the current active sink URI that has been configured for the
 	//   Source.
 	duckv1.SourceStatus `json:",inline"`
+
+	// Total number of consumers targeted by this source
+	// +optional
+	// For round-tripping only.
+	Consumers int32 `json:"consumers,omitempty"`
```
Comment (slinkydeveloper) [inline pkg/apis/sources/v1alpha1/kafka_types.go:139]: Why do we need this? Isn't the one in the spec enough?

Comment (lionelvillard) [reply pkg/apis/sources/v1alpha1/kafka_types.go:139]: yes it's a required field for enabling the scale subresource. See https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource

File: pkg/apis/sources/v1alpha1/kafka_types.go (hunk 4 above)
Comment (lionelvillard) [reply pkg/apis/sources/v1alpha1/kafka_types.go:102]: vaguer is a better :-) More consumers = faster, that's the only thing we need to know at the level.

Comment (slinkydeveloper) [reply pkg/apis/sources/v1alpha1/kafka_types.go]: +1 for the wording you just committed

**Comment by lionelvillard** [comment]
@slinkydeveloper @matzew can we merge this one now?

**Comment by slinkydeveloper** [comment]
Fine for me

/lgtm
/approve

**Comment by lionelvillard** [comment]
only fine?! :-)

**Pull request merged**